File: src/cliente/cliente.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"net"
	"time"
)

type Message struct {
	Type      string `json:"type"`
	ClientID  string `json:"client_id"`
	Timestamp int64  `json:"timestamp"`
	Sequence  int    `json:"sequence"`
	Data      string `json:"data"`
}

type Response struct {
	Type      string `json:"type"`
	Status    string `json:"status"`
	Timestamp int64  `json:"timestamp"`
	Message   string `json:"message"`
}

func conectar(host string, porta int, numMensagens int, clienteID string) error {
	address := fmt.Sprintf("%s:%d", host, porta)
	
	conn, err := net.Dial("tcp", address)
	if err != nil {
		return err
	}
	defer conn.Close()

	// Enviar múltiplas mensagens conforme configurado
	for i := 1; i <= numMensagens; i++ {
		msg := Message{
			Type:      "PING",
			ClientID:  clienteID,
			Timestamp: time.Now().UnixNano() / int64(time.Millisecond),
			Sequence:  i,
			Data:      fmt.Sprintf("PING %d do cliente %s", i, clienteID),
		}

		msgBytes, err := json.Marshal(msg)
		if err != nil {
			return err
		}

		// Enviar tamanho da mensagem primeiro (protocolo)
		msgSize := fmt.Sprintf("%08d", len(msgBytes))
		_, err = conn.Write([]byte(msgSize))
		if err != nil {
			return err
		}

		// Enviar a mensagem
		_, err = conn.Write(msgBytes)
		if err != nil {
			return err
		}

		// Ler resposta
		sizeBuffer := make([]byte, 8)
		_, err = conn.Read(sizeBuffer)
		if err != nil {
			return err
		}

		var responseSize int
		fmt.Sscanf(string(sizeBuffer), "%d", &responseSize)

		responseBuffer := make([]byte, responseSize)
		_, err = conn.Read(responseBuffer)
		if err != nil {
			return err
		}

		var response Response
		err = json.Unmarshal(responseBuffer, &response)
		if err != nil {
			return err
		}

		// Pequena pausa entre mensagens para evitar sobrecarga
		if i < numMensagens {
			time.Sleep(10 * time.Millisecond)
		}
	}

	return nil
}

func main() {
	host := flag.String("host", "127.0.0.1", "Host do servidor")
	porta := flag.Int("porta", 8000, "Porta do servidor")
	numMensagens := flag.Int("mensagens", 1, "Número de mensagens a enviar")
	clienteID := flag.String("cliente-id", "", "ID único do cliente")
	flag.Parse()

	if *clienteID == "" {
		*clienteID = fmt.Sprintf("cliente_%d", time.Now().UnixNano())
	}

	err := conectar(*host, *porta, *numMensagens, *clienteID)
	if err != nil {
		// Falha silenciosa para não poluir os logs de teste
		return
	}
}

File: src/cliente (local)/cliente.py
import socket
import argparse
import json
import time

def conectar(host, porta, num_mensagens, cliente_id):
    try:
        cliente = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        cliente.connect((host, porta))
        
        # Enviar múltiplas mensagens conforme configurado
        for i in range(1, num_mensagens + 1):
            msg = {
                "type": "PING",
                "client_id": cliente_id,
                "timestamp": int(time.time() * 1000),
                "sequence": i,
                "data": f"PING {i} do cliente {cliente_id}"
            }
            
            msg_bytes = json.dumps(msg).encode('utf-8')
            
            # Enviar tamanho da mensagem primeiro (protocolo)
            msg_size = f"{len(msg_bytes):08d}".encode('utf-8')
            cliente.sendall(msg_size)
            
            # Enviar a mensagem
            cliente.sendall(msg_bytes)
            
            # Ler resposta
            size_buffer = cliente.recv(8)
            if len(size_buffer) < 8:
                break
                
            response_size = int(size_buffer.decode('utf-8'))
            response_buffer = cliente.recv(response_size)
            
            if len(response_buffer) < response_size:
                break
                
            response = json.loads(response_buffer.decode('utf-8'))
            
            # Pausa mínima entre mensagens (necessária para estabilidade do teste)
            if i < num_mensagens:
                time.sleep(0.005) 
        
        cliente.close()
    except Exception as e:
        pass

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--host", type=str, required=True)
    parser.add_argument("--porta", type=int, required=True)
    parser.add_argument("--mensagens", type=int, default=1, help="Número de mensagens a enviar")
    parser.add_argument("--cliente-id", type=str, default="", help="ID único do cliente")
    args = parser.parse_args()

    if not args.cliente_id:
        args.cliente_id = f"cliente_{int(time.time() * 1000000)}"

    conectar(args.host, args.porta, args.mensagens, args.cliente_id)
